### PR TITLE
Make SYMENGINE_ASSERT use custom assert instead of assert()

### PR DESCRIPTION
**Pull request by Sumith1896**
En route to fixing https://github.com/sympy/symengine/issues/482
/cc @certik @isuruf

### CODE REVIEW

**Comment by certik** [comment]
Why do you use do/while? Wouldn't a simple block suffice?

**Comment by Sumith1896** [comment]
@certik 
If I remove the `do-while`, just `if ` remains
Say I do
```
if(X)
SYMENGINE_ASSERT(Y)
else
something_else;
```
The `else` is matched to the `if` of `SYMENGINE_ASSERT`.
This practice I noted in places, here  bytes.com/topic/c/answers/215019-use-do-while-0-assert-macro not something I knew of beforehand.
Let me know what you think.

**Comment by certik** [comment]
Cannot you use the {} block? Then the else cannot be match to an if inside
the block.

**Comment by Sumith1896** [comment]
Oh I just noticed that we don't put semi-colon after usage of macros like `SYMENGINE_ASSERT` etc.
By standard practice, people usually put `;` after, which means
```
#define FOO(x) do { foo(x); bar(x); } while (0)
```
serves the syntactic purpose while just braces `{}` don't.
I think for us braces would do the job.
Changing now.

**Comment by Sumith1896** [comment]
@certik Done.

**Comment by certik** [comment]
Ah I see what you mean. Either is fine with me. Just using {} seems simpler
to me.

**Comment by Sumith1896** [comment]
Travis passes.

File: symengine/symengine_assert.h [hunk 1]
```diff
@@ -1,10 +1,21 @@
...
+// SYMENGINE_ASSERT uses internal functions to perform as assert
+// so that there is no effect with NDEBUG
 #if !defined(SYMENGINE_ASSERT)
 #if defined(WITH_SYMENGINE_ASSERT)
-#define SYMENGINE_ASSERT(cond) assert(cond);
+#define stringize(s) #s
+#define XSTR(s) stringize(s)
+#define SYMENGINE_ASSERT(cond) \
+{ \
+if (0 == (cond)) { \
```
Comment (certik) [inline symengine/symengine_assert.h]: Is there a reason why this cannot be written as `if (!(cond)) { \` ? Not a big deal, just asking.

Comment (Sumith1896) [inline symengine/symengine_assert.h]: The standard `assert()` takes in int: http://en.cppreference.com/w/cpp/error/assert
ans hence this practice is adopted and used in the internal `assert()` too. I am not finding the exact place where I read this.

Comment (certik) [inline symengine/symengine_assert.h]: I see. That's fine.

**Comment by certik** [comment]
The shippable bugs seem internal to Shippable. Merging.

**Comment by Sumith1896** [comment]
Thanks for merging!

**Pull request merged**